src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs

```diff
@@ -53,51 +55,53 @@ public IAsyncToken BeginAsyncOperation(string name, object tag = null, [CallerFi
 
         private void Increment()
         {
-            lock (_gate)
-            {
-                _counter++;
-            }
+            Contract.ThrowIfFalse(_gate.LockHeldByMe());
+            _counter++;
         }
 
         private void Decrement(AsyncToken token)
         {
-            lock (_gate)
+            Contract.ThrowIfFalse(_gate.LockHeldByMe());
+
+            _counter--;
+            if (_counter == 0)
             {
-                _counter--;
-                if (_counter == 0)
+                foreach (var task in _pendingTasks)
                 {
-                    foreach (var task in _pendingTasks)
-                    {
-                        task.SetResult(true);
-                    }
-
-                    _pendingTasks.Clear();
+                    // setting result of a task can cause await machinary to wake up awaited code
+                    // and run code inline. that basically means random code running at the same thread
+                    // as the thread SetResult is called. so make sure we do in another thread (basically outside of the lock).
+                    // also, to prevent re-enterance bug, use NonReentrantLock to explicitly block
+                    // re-enterance
+                    Task.Run(() => task.SetResult(true));
```

Comment:
TaskCompletionSource now can be passed TaskCreationOptions.RunContinuationsAsynchronously to the constructor, which does the behavior you want. Just do that and then you can remove the comment entirely.